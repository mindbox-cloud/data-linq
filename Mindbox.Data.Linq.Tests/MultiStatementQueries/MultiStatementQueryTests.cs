using System.Data;
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Snapshooter.MSTest;

namespace Mindbox.Data.Linq.Tests.MultiStatementQueries;

[TestClass]
public class MultiStatementQueryTests
{
    [TestMethod]
    public void Translate_NoFilter_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().AsQueryable().Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TopLevelFilterAgainstConstant_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c2 => c2.TempPasswordEmail == "123").Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_ChainedWhere_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c2 => c2.TempPasswordEmail == "123").Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TopLevelFilterAgainstVariable_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var someEmail = "123";
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => c.TempPasswordEmail == someEmail).Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TopLevelFilterAgainstBooleanField_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => c.IsDeleted).Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TopLevelFilterAgainstNotBooleanField_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => !c.IsDeleted).Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TopLevelMultipleSimleFilters_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var someEmail = "123";
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => c.TempPasswordEmail == someEmail && c.Id > 10 && !c.IsDeleted).Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TopLevelMultipleSimleFiltersOnSeveralWhereBlocks_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var someEmail = "123";
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => c.TempPasswordEmail == someEmail).Where(c => c.Id > 10).Where(c => c.Id > 10 && !c.IsDeleted).Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TableLinkedViaReferenceAssociation_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => c.Area.Name == "SomeArea").Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TableLinkedViaReferenceChainAssociation_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>().Where(c => c.Area.SubArea.Name == "SomeSubArea").Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TableLinkedViaReferenceChainAssociationAndFilterOnAllAssociations_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>()
            .Where(c => c.Area.Name == "SomeArea")
            .Where(c => c.Area.SubArea.Name == "SomeSubArea").Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }

    [TestMethod]
    public void Translate_TableJoinByDataFieldViaWhere_Success()
    {
        // Arrange
        using var contextAndConnection = new DataContextAndConnection();

        // Act
        var customerActions = contextAndConnection.DataContext.GetTable<CustomerAction>();
        var queryExpression = contextAndConnection.DataContext
            .GetTable<Customer>()
            .Where(c => customerActions.Where(ca => ca.CustomerId == ca.Id).Any(ca => ca.ActionTemplateId == 10))
            .Expression;
        var query = SqlQueryTranslator.Transalate(queryExpression, new DbColumnTypeProvider());

        // Assert
        query.CommandText.MatchSnapshot();
    }
    
    // Nested joint
    // Querable join
    // Select with anonympus types
}