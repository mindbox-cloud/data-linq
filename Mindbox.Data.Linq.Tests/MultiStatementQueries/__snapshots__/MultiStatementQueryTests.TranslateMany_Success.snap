****************************** Original expression A **********************************
Table(Customer).Where(c => value(Mindbox.Data.Linq.Tests.MultiStatementQueries.MultiStatementQueryTests+<>c__DisplayClass22_0).orders.Where(o => (o.CurrentCustomer == c)).SelectMany(o => o.History.Single(hi => (hi.IsCurrentOtherwiseNull != null)).Purchases).Where(p => (((p.PriceForCustomerOfLine / p.Count) != null) AndAlso ((p.PriceForCustomerOfLine / p.Count) >= Convert(Convert(123, Decimal), Nullable`1)))).Any())

****************************** Original expression B **********************************
Table(Customer).Where(c => c.CustomFieldValues.Any(f => (f.FieldValue == "ASDF")))

****************************** SQL result *********************************************
DECLARE @tabledirectcrm_Customers TABLE(
    Id int not null
)
DECLARE @tabledirectcrm_RetailOrders TABLE(
    CustomerId int null,
    Id int not null
)
DECLARE @tabledirectcrm_RetailOrderHistoryItems TABLE(
    Id int not null,
    IsCurrentOtherwiseNull bit null,
    RetailOrderId int not null
)
DECLARE @tabledirectcrm_RetailOrderPurchases TABLE(
    Count decimal(18,2) not null,
    PriceForCustomerOfLine decimal(18,2) not null,
    RetailOrderHistoryItemId bigint not null
)
DECLARE @tabledirectcrm_CustomerCustomFieldValues TABLE(
    CustomerId int not null,
    FieldValue nvarchar(32) not null
)

INSERT INTO @tabledirectcrm_Customers(Id)
    SELECT current.Id
        FROM directcrm.Customers AS current
            WHERE Id = @__id
SELECT * FROM @tabledirectcrm_Customers

INSERT INTO @tabledirectcrm_RetailOrders(CustomerId, Id)
    SELECT current.CustomerId, current.Id
        FROM directcrm.RetailOrders AS current
            INNER JOIN @tabledirectcrm_Customers AS previous ON
                previous.Id = current.CustomerId
SELECT * FROM @tabledirectcrm_RetailOrders

INSERT INTO @tabledirectcrm_RetailOrderHistoryItems(Id, IsCurrentOtherwiseNull, RetailOrderId)
    SELECT current.Id, current.IsCurrentOtherwiseNull, current.RetailOrderId
        FROM directcrm.RetailOrderHistoryItems AS current
            INNER JOIN @tabledirectcrm_RetailOrders AS previous ON
                previous.Id = current.RetailOrderId
SELECT * FROM @tabledirectcrm_RetailOrderHistoryItems

INSERT INTO @tabledirectcrm_RetailOrderPurchases(Count, PriceForCustomerOfLine, RetailOrderHistoryItemId)
    SELECT current.Count, current.PriceForCustomerOfLine, current.RetailOrderHistoryItemId
        FROM directcrm.RetailOrderPurchases AS current
            INNER JOIN @tabledirectcrm_RetailOrderHistoryItems AS previous ON
                previous.Id = current.RetailOrderHistoryItemId
SELECT * FROM @tabledirectcrm_RetailOrderPurchases

INSERT INTO @tabledirectcrm_CustomerCustomFieldValues(CustomerId, FieldValue)
    SELECT current.CustomerId, current.FieldValue
        FROM directcrm.CustomerCustomFieldValues AS current
            INNER JOIN @tabledirectcrm_Customers AS previous ON
                previous.Id = current.CustomerId
SELECT * FROM @tabledirectcrm_CustomerCustomFieldValues

****************************** Rewritten expression A *********************************
resultSet => resultSet.GetTable("directcrm.Customers").Where(c => resultSet.GetTable("directcrm.RetailOrders").Where(o => (o.GetReferencedRow("directcrm.Customers", "CustomerId", "Id") == c)).SelectMany(o => o.GetReferencedRowsArray("directcrm.RetailOrderHistoryItems", "Id", "RetailOrderId").Single(hi => (hi.GetValue("IsCurrentOtherwiseNull") != null)).GetReferencedRowsArray("directcrm.RetailOrderPurchases", "Id", "RetailOrderHistoryItemId")).Where(p => (((p.GetValue("PriceForCustomerOfLine") / p.GetValue("Count")) != null) AndAlso ((p.GetValue("PriceForCustomerOfLine") / p.GetValue("Count")) >= Convert(Convert(123, Decimal), Nullable`1)))).Any()).Any()

****************************** Rewritten expression B *********************************
resultSet => resultSet.GetTable("directcrm.Customers").Where(c => c.GetReferencedRowsArray("directcrm.CustomerCustomFieldValues", "Id", "CustomerId").Any(f => (f.GetValue("FieldValue") == "ASDF"))).Any()

